include/seqan3/alignment/pairwise/alignment_selector.hpp

```diff
@@ -0,0 +1,145 @@
+                return align_result<type_list<uint32_t,
+                                              score_type,
+                                              std::pair<size_t, size_t>,
+                                              std::pair<size_t, size_t>>>{};
+            else if constexpr (get<align_cfg::id::output>(configuration_t{}) ==
+                               align_result_key::trace)
+                return align_result<type_list<uint32_t,
+                                              score_type,
+                                              std::pair<size_t, size_t>,
+                                              std::pair<size_t, size_t>,
+                                              std::tuple<std::vector<gapped<seq1_value_type>>,
+                                                         std::vector<gapped<seq2_value_type>>>>>{};
+            else
+                return align_result<type_list<uint32_t,
+                                              score_type>>{};
+        }
+        else
+        {
+            return align_result<type_list<uint32_t, score_type>>{};
+        }
+    }
+
+    //!\brief The determined result type.
+    using type = decltype(_determine());
+};
+
+/*!\brief Selects the correct alignment algorithm based on the algorithm configuration.
+ * \ingroup pairwise
+ * \tparam seq_tuple_t     A tuple like object containing the two source sequences.
+ *                         Must model seqan3::tuple_like_concept.
+ * \tparam configuration_t The specified configuration type.
+ */
+template <tuple_like_concept seq_tuple_t, typename configuration_t>
+struct alignment_selector
+{
+    //!\brief The configuration stored globally for all alignment instances.
+    configuration_t config;
+
+    //!\brief The result type of invoking the algorithm.
+    using result_type = typename determine_result_type<std::tuple_element_t<0, std::remove_reference_t<seq_tuple_t>>,
+                                                       std::tuple_element_t<1, std::remove_reference_t<seq_tuple_t>>,
+                                                       configuration_t>::type;
+
+    /*!\brief Selects the corresponding alignment algorithm based on compile time and runtime decisions.
+     * \param[in] seq The sequences as a tuple.
+     * \returns A std::function object to be called within the alignment execution.
+     *
+     * \details
+     *
+     * \todo Write detail description and explain rationale about the function object.
+     */
+    template <tuple_like_concept _seq_tuple_t>
+    auto select(_seq_tuple_t && seq)
+    {
+        //TODO Currently we only support edit_distance. We need would actually need real checks for this.
+        std::function<result_type(result_type &)> func =
+            pairwise_alignment_edit_distance_unbanded{std::get<0>(std::forward<_seq_tuple_t>(seq)),
+                                                      std::get<1>(std::forward<_seq_tuple_t>(seq)),
```

Comment:
interesting question! I mean of i forward the tuple as a r-value to the get interface, it will move out the contained element at the index but leave the other indices still valid. Since it is a reference, the tuple is not moved "away", so in the second invocation I move out the second sequence. 🤔

Reply:
if you move something into a function it is gone.

Reply:
No, that's not true in general. std::move and std::forward don't move anything, they just cast things to being an rvalue reference. This signals to a function that it *can* take the whole thing which *can* happen e.g. if the function takes its parameter by value. But it doesn't have to, e.g. std::get doesn't do this, it takes by lvalue-reference or revalued reference and then returns lvalue or revalued references to.the contained elements. Nothing is invalidated anywhere.

Reply:
Ok, unfortunately everyone is right.
Forwarding an rvalue should not invoke the move constructor of the class but simply forward the rvalue (by casting to ravalue). For std::get and STL tuples this is not a problem, as the function does not change the object but would move out the respective element. However, I think since c++17 I need to use `get` as an extension point really (`using std::get;` and then use unnamed get to allow ADL), allowing other tuple-like implementations. Now,  here I have no guarantee that the passed object is not modified somehow because the function could, and using it again afterwards could lead to unintended behaviour. Of course that would be shitty implementation but who knows, what the intentions were.
So if I am not mistaken I should change it to, which should be well defined for every tuple like obecjt:

```cpp
...
auto && [seq1, seq2] = std::forward<tpl_t>(tpl);
auto function = ...{std::forward<decltype(seq1)>(seq1), std::forward<decltype(seq2)>(seq2)};
...
```

Anyway, I am confused as hell and will test it thorough with all combinations of l and rvalue states.